### PR TITLE
Allow cross validation with 'bring your own' Lightning models (without ensemble building)

### PR DESCRIPTION
**Pull request by dumbledad**
When complete, will be closing #476 

- [x] Ensure that your PR is small, and implements one change.
- [x] Add unit tests for all functions that you introduced or modified.
- [ ] ~~Run PyCharm's code cleanup tools on your Python files.~~ (I'm using VSCode)
- [x] Link the correct GitHub issue for tracking.
- [x] Update the [Changelog](CHANGELOG.md) file: Describe your change in terms of 
Added/Changed/Removed/... in the "Upcoming" section.
- [x] When merging your PR, replace the default merge message with a description of your PR,
and if needed a motivation why that change was required.

### CODE REVIEW

File: InnerEye/ML/run_ml.py [hunk 1]
```diff
@@ -242,14 +242,10 @@ def is_offline_run(self) -> bool:
         return is_offline_run_context(RUN_CONTEXT)
 
     @property
-    def innereye_config(self) -> DeepLearningConfig:
+    def innereye_config(self) -> Optional[DeepLearningConfig]:
         """
-        Gets the model configuration object for all built-in InnerEye models. Raises an exception if the present
-        object trains a LightningContainer that is not a built-in InnerEye model.
+        Gets the model configuration object for all built-in InnerEye models.
```
Comment (ant0nsc) [inline InnerEye/ML/run_ml.py:247]: is that change really necessary anymore?

**Review by ant0nsc — CHANGES_REQUESTED** [review]
Looking good so far, but please increase test coverage. At the moment, there is no clear proof in the tests that submitting an AML job with a container model and crossval would actually work.

**Comment by dumbledad** [comment]
> Looking good so far, but please increase test coverage. At the moment, there is no clear proof in the tests that submitting an AML job with a container model and crossval would actually work.

That's a good idea. I thought we had agreed to use unit tests that did not actually run an AML experiment so that the test did not incur a time penalty and a cost. But I think an end-to-end test with known outputs from a known model is a safer bet. I'll do that.

File: InnerEye/ML/run_ml.py (hunk 1 above)
Comment (dumbledad) [reply InnerEye/ML/run_ml.py]: No I don't think we do need that, you are correct. Fixed

File: InnerEye/ML/deep_learning_config.py [hunk 2]
```diff
@@ -276,6 +280,52 @@ def get_effective_random_seed(self) -> int:
...
+        """
+        Returns the total number of HyperDrive/offline runs required to sample the entire
+        cross validation parameter space.
+        """
+        return self.number_of_cross_validation_splits
+
+    @abstractmethod
+    def get_cross_validation_hyperdrive_config(self, run_config: ScriptRunConfig) -> HyperDriveConfig:
+        """
+        Is implemented in both ModelConfigBase and LightningContainer
```
[the remark this answers is not in the document]
Comment (dumbledad) [reply InnerEye/ML/deep_learning_config.py]: Fixed.

File: InnerEye/ML/deep_learning_config.py [hunk 3]
```diff
@@ -276,6 +280,52 @@ def get_effective_random_seed(self) -> int:
...
+        return GridParameterSampling(parameter_space={
+            CROSS_VALIDATION_SPLIT_INDEX_TAG_KEY: choice(list(range(self.number_of_cross_validation_splits))),
+        })
+
+    def get_total_number_of_cross_validation_runs(self) -> int:
+        """
+        Returns the total number of HyperDrive/offline runs required to sample the entire
+        cross validation parameter space.
+        """
+        return self.number_of_cross_validation_splits
```
[the remark this answers is not in the document]
Comment (dumbledad) [reply InnerEye/ML/deep_learning_config.py]: Done.

File: InnerEye/ML/configs/other/HelloContainer.py [hunk 4]
```diff
@@ -30,15 +30,22 @@ class HelloDataset(Dataset):
...
+            root_folder: Path,
+            start_index: int,
+            end_index: int,
+            raw_data: List = None) -> None:
         """
         Creates the 1-dim regression dataset.
         :param root_folder: The folder in which the data file lives ("hellocontainer.csv")
         :param start_index: The first row to read.
         :param end_index: The last row to read (exclusive)
+        :param raw_data: (Optional, default None) if supplying the raw data, e.g. from a cross validation split
```
Comment (ant0nsc) [inline InnerEye/ML/configs/other/HelloContainer.py:44]: I don't get that docstring.

File: InnerEye/ML/configs/other/HelloContainer.py [hunk 5]
```diff
@@ -47,16 +54,68 @@ def __len__(self) -> int:
...
+            split_raw_data_remaining = np.array_split(raw_data_remaining, number_of_cross_validation_splits)
+            self.val = HelloDataset.from_raw_data(split_raw_data_remaining[cross_validation_split_index])
+            train: List[Any] = []
+            for train_index in range(number_of_cross_validation_splits):
+                if train_index != cross_validation_split_index:
+                    if len(train) > 0:
+                        train = np.append(train, split_raw_data_remaining[train_index], axis=0)
+                    else:
+                        train = split_raw_data_remaining[train_index]
+            self.train = HelloDataset.from_raw_data(train)
```
Comment (ant0nsc) [inline InnerEye/ML/configs/other/HelloContainer.py:112]: Can we use a standard library function for that, like https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html?

File: InnerEye/ML/lightning_container.py [hunk 6]
```diff
@@ -288,6 +294,26 @@ def create_lightning_module_and_store(self) -> None:
...
+            primary_metric_goal=PrimaryMetricGoal.MINIMIZE,
+            max_total_runs=self.number_of_cross_validation_splits
+        )
+
+    def get_parameter_search_hyperdrive_config(self, _: ScriptRunConfig) -> HyperDriveConfig:  # type: ignore
+        """
+        Parameter search is not implemented. It should be implemented in a sub class if needed.
+        """
+        raise NotImplementedError("Parameter search is not implemented. It should be implemented in a sub class if needed.")
+
```
Comment (ant0nsc) [inline InnerEye/ML/lightning_container.py:316]: I don't think that this is needed, because the base class method is marked as abstract. If you wish to point readers to the possibility of that being changed, then please move it further up, see the comment in line 275

File: InnerEye/ML/deep_learning_config.py [hunk 7]
```diff
@@ -283,6 +287,51 @@ def get_effective_random_seed(self) -> int:
...
+        Turns the given AzureML estimator (settings for running a job in AzureML) into a configuration object
+        for doing hyperparameter searches.
+
+        This abstract method is implemented in ModelConfigBase but not in LightningContainer. Classes inheriting
+        from LightningCountainer will wish to provide an implementation if they provide hyper drive support for
+        parameter search.
+        :param run_config: AzureML estimator
+        :return: HyperDriveConfigs
+        """
+        pass
```
Comment (ant0nsc) [inline InnerEye/ML/deep_learning_config.py:333]: Thinking about that again... I think that's at the wrong place of the class hierarchy. It would be better if that lived at LightningContainer level. InnerEye container could implement that by falling back to model_config.get_parameter_search.... This would save you also one switch statement in your code (runner.py line 224)

File: Tests/ML/configs/other/test_hello_container.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    or requesting a number of splits larger than the available data).
+     - We test that between splits the train and validation data differ, and the test data remains the same.
+     - We test that all the data is used in each split.
+     - We test that across all the splits the validation data use all of the non-test data.
+    '''
+    # Get full data-set for comparison
+    root_folder = fixed_paths_for_tests.full_ml_test_data_path()
+    data_module_no_xval = HelloDataModule(root_folder=root_folder)
+
+    for number_of_cross_validation_splits in [0, 1, 5]:
```
Comment (ant0nsc) [inline Tests/ML/configs/other/test_hello_container.py:26]: nitpicking - I think it should also fail for splits=1. Or are you treating that the same as splits=0?

**Review by ant0nsc — CHANGES_REQUESTED** [review]

File: Tests/ML/configs/other/test_hello_container.py (hunk 8 above)
Comment (dumbledad) [reply Tests/ML/configs/other/test_hello_container.py:26]: I'm treating it the same as 0, but I can swap it to fail if that's what ML people would expect @ant0nsc?

File: InnerEye/ML/configs/other/HelloContainer.py (hunk 4 above)
Comment (dumbledad) [reply InnerEye/ML/configs/other/HelloContainer.py:44]: I don't like the two ways this class is constructed, in another language it would have two constructors, but we cannot do that in Python. I'll refactor.

Comment (dumbledad) [reply InnerEye/ML/configs/other/HelloContainer.py]: Done

File: InnerEye/ML/configs/other/HelloContainer.py (hunk 5 above)
Comment (dumbledad) [reply InnerEye/ML/configs/other/HelloContainer.py]: Done.

File: InnerEye/ML/lightning_container.py (hunk 6 above)
Comment (dumbledad) [reply InnerEye/ML/lightning_container.py:316]: Removed.

File: InnerEye/ML/deep_learning_config.py (hunk 7 above)
Comment (dumbledad) [reply InnerEye/ML/deep_learning_config.py:333]: But it needs to be in a parent of both LightningContainer and DeepLEarningConfig. For example, it is overridden in InnerEye.configs.segmentation.HelleWorld which inherits from SegmentationModelBase -> ModelConfigBase -> DeepLearningConfig -> WorkflowParam. No?

File: Tests/ML/configs/other/test_hello_container.py (hunk 8 above)
Comment (ant0nsc) [reply Tests/ML/configs/other/test_hello_container.py:26]: That's fine, it's just a piece of example code

File: InnerEye/ML/deep_learning_config.py (hunk 7 above)
Comment (dumbledad) [reply InnerEye/ML/deep_learning_config.py]: Done, just waiting for tests to run. I just moved `get_parameter_search_hyperdrive_config` back to `ModelConfigBase` which does not save me the switch you mention in runner, as that is about a different (but similarly named) method `get_hyperdrive_config`. Did you want all of them moved?

Comment (dumbledad) [reply InnerEye/ML/deep_learning_config.py]: Done for all the methods.

**Review by ant0nsc — APPROVED** [review]

File: InnerEye/ML/lightning_container.py [hunk 9]
```diff
@@ -288,6 +301,40 @@ def create_lightning_module_and_store(self) -> None:
...
+        else:
+            return self.get_parameter_search_hyperdrive_config(run_config)
+
+    def get_cross_validation_hyperdrive_sampler(self) -> GridParameterSampling:
+        """
+        Returns the cross validation sampler, required to sample the entire parameter space for cross validation.
+        """
+        return GridParameterSampling(parameter_space={
+            CROSS_VALIDATION_SPLIT_INDEX_TAG_KEY: choice(list(range(self.number_of_cross_validation_splits))),
+        })
```
[the remark this answers is not in the document]
Comment (dumbledad) [reply InnerEye/ML/lightning_container.py:336]: We use it to make the HyperDriveConfig in the get_cross_validation_hyperdrive_config method. We could subsume it into that method, but the existing code in model_config_base does not, and so I was. just maintaining consistency. I've changed it now so that it is not a separate method.

File: InnerEye/ML/configs/other/HelloContainer.py [hunk 10]
```diff
@@ -30,16 +31,13 @@ class HelloDataset(Dataset):
     # x = torch.rand((N, 1)) * 10
     # y = 0.2 * x + 0.1 * torch.randn(x.size())
     # xy = torch.cat((x, y), dim=1)
-    # np.savetxt("InnerEye/ML/configs/other/hellocontainer.csv", xy.numpy(), delimiter=",")
-    def __init__(self, root_folder: Path, start_index: int, end_index: int) -> None:
+    # np.savetxt("Tests/ML/test_data/hellocontainer.csv", xy.numpy(), delimiter=",")
+    def __init__(self, raw_data: List) -> None:
```
[the remark this answers is not in the document]
Comment (dumbledad) [reply InnerEye/ML/configs/other/HelloContainer.py]: Done

**Review by ant0nsc — APPROVED** [review]

File: InnerEye/ML/configs/other/HelloContainer.py [hunk 11]
```diff
@@ -30,16 +31,15 @@ class HelloDataset(Dataset):
     # x = torch.rand((N, 1)) * 10
     # y = 0.2 * x + 0.1 * torch.randn(x.size())
     # xy = torch.cat((x, y), dim=1)
-    # np.savetxt("InnerEye/ML/configs/other/hellocontainer.csv", xy.numpy(), delimiter=",")
-    def __init__(self, root_folder: Path, start_index: int, end_index: int) -> None:
+    # np.savetxt("Tests/ML/test_data/hellocontainer.csv", xy.numpy(), delimiter=",")
```
Comment (Shruthi42) [inline InnerEye/ML/configs/other/HelloContainer.py:34]: hellocontainer.csv is now in InnerEye/ML/configs/other/

File: InnerEye/ML/lightning_container.py [hunk 12]
```diff
@@ -288,6 +301,35 @@ def create_lightning_module_and_store(self) -> None:
...
+        :param run_config: The AzureML run configuration object that training for an individual model.
+        :return: A hyperdrive configuration object.
+        """
+        return HyperDriveConfig(
+            run_config=run_config,
+            hyperparameter_sampling=GridParameterSampling(
+                parameter_space={
+                    CROSS_VALIDATION_SPLIT_INDEX_TAG_KEY: choice(list(range(self.number_of_cross_validation_splits)))
+                }),
+            primary_metric_name=TrackedMetrics.Val_Loss.value,
```
Comment (Shruthi42) [inline InnerEye/ML/lightning_container.py:316]: Is it up to the user to make sure they log a metric with the same name in their model configs?

File: InnerEye/ML/configs/other/HelloContainer.py (hunk 11 above)
Comment (dumbledad) [reply InnerEye/ML/configs/other/HelloContainer.py:34]: Good catch, Anton moved that as part of his fixes earlier this week. Ill fix that.

Comment (dumbledad) [reply InnerEye/ML/configs/other/HelloContainer.py]: Done

File: InnerEye/ML/lightning_container.py (hunk 12 above)
Comment (dumbledad) [reply InnerEye/ML/lightning_container.py:316]: That's interesting. I'm not sure, @ant0nsc?

Comment (dumbledad) [reply InnerEye/ML/lightning_container.py:319]: Done. I've added a comment in the method you highlighted and some directives in the BYOM documentation. Both point to the example in `HelloRegression(LightningModule)` where we provide the val/Loss logging.

**Review by Shruthi42 — APPROVED** [review]

**Review by ant0nsc — APPROVED** [review]
